processor/src/main/java/com/linecorp/decaton/processor/runtime/ExecutionScheduler.java

```diff
@@ -60,27 +59,45 @@ void sleep(long millis) throws InterruptedException {
     }
 
     // visible for testing
-    void waitOnScheduledTime(TaskMetadata metadata) throws InterruptedException {
-        long timeUntilProcess = metadata.scheduledTimeMillis() - currentTimeMillis.get();
-        if (timeUntilProcess > 0) {
-            if (logger.isDebugEnabled()) {
-                logger.debug("Sleeping {} ms awaiting scheduled time for task in {} - ({})",
-                             timeUntilProcess, scope, metadata);
-            }
-            sleep(timeUntilProcess);
+    long waitOnScheduledTime(TaskMetadata metadata) throws InterruptedException {
+        long timeToWaitMs = metadata.scheduledTimeMillis() - currentTimeMillis.get();
+        if (timeToWaitMs > 0) {
+            log.debug("Sleeping {} ms awaiting scheduled time for task in {} - ({})",
+                      timeToWaitMs, scope, metadata);
+            sleep(timeToWaitMs);
         }
-
-        metrics.tasksSchedulingDelay.record(Math.max(timeUntilProcess, 0), TimeUnit.MILLISECONDS);
+        return timeToWaitMs;
     }
 
+    /**
+     * Wait for the time ready to start executing a task which is associated with the given
+     * {@link TaskMetadata}.
+     * This method pauses until all below criteria tells it is ready to start executing next task.
+     * - Execution time configured in task metadata
+     * - Per-partition rate limiting
+     *
+     * This method might returns earlier than the time ready to execute the task when {@link #close()} called.
+     *
+     * @param metadata the {@link TaskMetadata} associated with the task waiting to be processed.
+     * @throws InterruptedException when interrupted.
+     */
     public void schedule(TaskMetadata metadata) throws InterruptedException {
         // For tasks which are configured to delay its execution.
-        waitOnScheduledTime(metadata);
+        long timeWaitedMs = waitOnScheduledTime(metadata);
+        if (terminated()) {
```

Comment:
Yes, because of:

1. It should not be recorded in below metric
2. It's pointless to enter rateLimiter which potentially blocks and is still unknown when exactly it will be closed despite we know the termination has began at this moment.